### PR TITLE
Spread & rest over intersection and unions

### PR DESCRIPTION
**Pull request by mhegazy**
Fixes https://github.com/Microsoft/TypeScript/issues/12520 , https://github.com/Microsoft/TypeScript/issues/12534 and  #12460

//CC @sandersn

### CODE REVIEW

File: tests/baselines/reference/restIntersectionOrIntersection.types [hunk 1]
```diff
@@ -0,0 +1,38 @@
+=== tests/cases/compiler/restIntersectionOrIntersection.ts ===
```
Comment (sandersn) [inline tests/baselines/reference/restIntersectionOrIntersection.types]: should be named restIntersectionOrUnion.ts, right?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -6102,11 +6108,37 @@ namespace ts {
...
+                left = emptyObjectType;
+            }
+
+            if (right.flags & TypeFlags.Union) {
+                const types = filter((<UnionType>right).types, t => !(t.flags & TypeFlags.Nullable));
+                if (types.length) {
+                    return getUnionType(map(types, t => getSpreadType(left, t, isFromObjectLiteral)));
+                }
+                else {
+                    right = emptyObjectType
```
Comment (sandersn) [inline src/compiler/checker.ts]: missing semicolon

File: tests/baselines/reference/objectSpreadNegative.errors.txt [hunk 3]
```diff
@@ -68,11 +66,7 @@ tests/cases/conformance/types/spread/objectSpreadNegative.ts(61,14): error TS269
     
     // null, undefined and primitives are not allowed
```
Comment (sandersn) [inline tests/baselines/reference/objectSpreadNegative.errors.txt]: can you update this comment?

**Comment by mhegazy** [comment]
@ahejlsberg can you take a look

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -6102,11 +6108,37 @@ namespace ts {
...
          * and right = the new element to be spread.
          */
-        function getSpreadType(left: Type, right: Type, isFromObjectLiteral: boolean): ResolvedType | IntrinsicType {
-            Debug.assert(!!(left.flags & (TypeFlags.Object | TypeFlags.Any)) && !!(right.flags & (TypeFlags.Object | TypeFlags.Any)), "Only object types may be spread.");
+        function getSpreadType(left: Type, right: Type, isFromObjectLiteral: boolean): Type  {
             if (left.flags & TypeFlags.Any || right.flags & TypeFlags.Any) {
                 return anyType;
             }
+
+            if (left.flags & TypeFlags.Union) {
```
Comment (vladima) [inline src/compiler/checker.ts]: nit: these two parts of code for left and right looks quite similar, can they be extracted to a function?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -6097,16 +6103,46 @@ namespace ts {
...
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: You can replace all of these new lines with just this:

```ts
            left = filterType(left, t => !(t.flags & TypeFlags.Nullable));
            if (left.flags & TypeFlags.Never) {
                return right;
            }
            right = filterType(right, t => !(t.flags & TypeFlags.Nullable));
            if (right.flags & TypeFlags.Never) {
                return left;
            }
            if (left.flags & TypeFlags.Union) {
                return mapType(left, t => getSpreadType(t, right, isFromObjectLiteral));
            }
            if (right.flags & TypeFlags.Union) {
                return mapType(right, t => getSpreadType(left, t, isFromObjectLiteral));
            }
```

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -11516,6 +11552,24 @@ namespace ts {
             }
        }
 
+        function isValidSpreadType(type: Type): boolean {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: The functional style is a lot shorter:

```ts
function isValidSpreadType(type: Type): boolean {
    return !!(type.flags & (TypeFlags.Any | TypeFlags.Null | TypeFlags.Undefined) ||
        type.flags & TypeFlags.Object && !isGenericMappedType(type) ||
        type.flags & TypeFlags.UnionOrIntersection && !forEach((<UnionOrIntersectionType>type).types, t => !isValidSpreadType(t)));
}
```

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -6097,16 +6103,46 @@ namespace ts {
             return symbol ? getLocalTypeParametersOfClassOrInterfaceOrTypeAlias(symbol) : undefined;
         }
 
+        function filterNulableTypes(union: UnionType): Type[] {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Spelling. Not sure you need to keep this function, but if you do, have it operate on any kind of type in the same way as `mapType` and `filterType`.

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -11516,6 +11538,24 @@ namespace ts {
             }
        }
 
+        function isValidSpreadType(type: Type): boolean {
+            if (type.flags & (TypeFlags.Any | TypeFlags.Null | TypeFlags.Undefined)) {
+                return true;
```
Comment (Kovensky) [inline src/compiler/checker.ts]: IIUC this should also fix #12460

File: src/compiler/checker.ts (hunk 6 above)
Comment (mhegazy) [inline src/compiler/checker.ts]: thanks. fixed.

**Comment by ahejlsberg** [comment]
:+1

**Pull request merged**